homeassistant/components/automation/__init__.py

```diff
@@ -485,35 +551,33 @@ def if_action(variables=None):
         """AND all conditions."""
         return all(check(hass, variables) for check in checks)
 
-    return if_action
+    if_action.config = if_configs
 
+    return if_action
 
-async def _async_process_trigger(hass, config, trigger_configs, name, action):
-    """Set up the triggers.
 
-    This method is a coroutine.
-    """
-    removes = []
-    info = {"name": name}
+@callback
+def _trigger_extract_device(trigger_conf: dict) -> Optional[str]:
+    """Extract devices from a trigger config."""
+    if trigger_conf[CONF_PLATFORM] != "device":
+        return None
 
-    for conf in trigger_configs:
-        platform = importlib.import_module(".{}".format(conf[CONF_PLATFORM]), __name__)
+    return trigger_conf[CONF_DEVICE_ID]
 
-        remove = await platform.async_attach_trigger(hass, conf, action, info)
 
-        if not remove:
-            _LOGGER.error("Error setting up trigger %s", name)
-            continue
+@callback
+def _trigger_extract_entities(trigger_conf: dict) -> List[str]:
+    """Extract entities from a trigger config."""
+    if trigger_conf[CONF_PLATFORM] in ("state", "numeric_state"):
+        return trigger_conf[CONF_ENTITY_ID]
 
-        _LOGGER.info("Initialized trigger %s", name)
-        removes.append(remove)
+    if trigger_conf[CONF_PLATFORM] == "zone":
+        return trigger_conf[CONF_ENTITY_ID] + [trigger_conf[CONF_ZONE]]
 
-    if not removes:
-        return None
+    if trigger_conf[CONF_PLATFORM] == "geo_location":
```

Comment:
Do we really want to add foreign domain logic to this domain. That's what we usually try to avoid, right?

Can we have the other domains register their logic on the automation domain instead?

Reply:
It's a file inside the automation integration. It's not ideal but thats how this integration currently works.

Reply:
This PR is now blocking the release, as discussed, we'll continue with the merge.

The further consensus of resolving this, can be done in a future followup PR.